File: 0670_maximum_swap/solution.cpp
// LeetCode 0670 - Maximum Swap
// https://leetcode.com/problems/maximum-swap/

#include <string>
#include <vector>

class Solution {
public:
    int maximumSwap(int num) {
        std::string digits = std::to_string(num);
        std::vector<int> last(10, -1);
        for (int i = 0; i < static_cast<int>(digits.size()); ++i) {
            last[digits[i] - '0'] = i;
        }
        for (int i = 0; i < static_cast<int>(digits.size()); ++i) {
            for (int candidate = 9; candidate > digits[i] - '0'; --candidate) {
                if (last[candidate] > i) {
                    std::swap(digits[i], digits[last[candidate]]);
                    return std::stoi(digits);
                }
            }
        }
        return num;
    }
};

File: 0670_maximum_swap/Solution.swift
// LeetCode 0670 - Maximum Swap
// https://leetcode.com/problems/maximum-swap/

class Solution {
    func maximumSwap(_ num: Int) -> Int {
        var digits = Array(String(num))
        var last = Array(repeating: -1, count: 10)
        for i in 0..<digits.count {
            last[Int(String(digits[i]))!] = i
        }
        for i in 0..<digits.count {
            let d = Int(String(digits[i]))!
            if d + 1 <= 9 {
                for candidate in stride(from: 9, through: d + 1, by: -1) {
                    if last[candidate] > i {
                        digits.swapAt(i, last[candidate])
                        return Int(String(digits))!
                    }
                }
            }
        }
        return num
    }
}
